### PR TITLE
W7 - hieucap - unit test

### PR DESCRIPTION
**Pull request by at-hieucap**
#### Related issues (Optional)
Task #84 

#### Description (Required)
- [x] Create screen
- [x] Create unit test

### CODE REVIEW

File: app/src/testAt-hieucapDebug/java/passWordUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import vn.asiantech.internship.unit_test.PasswordValidation;
+
+import static org.mockito.Mockito.when;
+
+/**
+ * Test validation password
+ * Created by tiboo on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class passWordUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/passWordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/passWordUnitTest.java]: ERROR: [ClassNamingConventions] Class names should begin with an uppercase character
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#ClassNamingConventions

File: app/src/testAt-hieucapDebug/java/passWordUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Mock
+    private Login mLogIn;
+
+    @Test
+    public void testDifferentUserName() {
+        when(mLogIn.getPassword()).thenReturn("dsad7ád7sad");
```
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/passWordUnitTest.java]: WARNING: [AvoidDuplicateLiterals] The String literal "dsad7ád7sad" appears 5 times in this file; the first occurrence is on line 24
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

File: app/src/testAt-hieucapDebug/java/userUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import vn.asiantech.internship.unit_test.UserNameValidation;
+
+import static org.mockito.Mockito.when;
+
+/**
+ * Test user name
+ * Created by tiboo on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class userUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/userUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/userUnitTest.java]: ERROR: [ClassNamingConventions] Class names should begin with an uppercase character
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#ClassNamingConventions

File: app/src/testAt-hieucapDebug/java/PasswordUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import vn.asiantech.internship.unit_test.PasswordValidation;
+
+import static org.mockito.Mockito.when;
+
+/**
+ * Test validation password
+ * Created by tiboo on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PasswordUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/PasswordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-hieucapDebug/java/UserUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import vn.asiantech.internship.unit_test.UserNameValidation;
+
+import static org.mockito.Mockito.when;
+
+/**
+ * Test user name
+ * Created by tiboo on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class UserUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/UserUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

**Comment by at-hieucap** [comment]
@at-trungnguyen  I fished exercise , please review for me,
Thank you so much !!

File: app/src/testAt-hieucapDebug/java/PasswordUnitTest.java (hunk 4 above)
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/PasswordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-hieucapDebug/java/UserUnitTest.java (hunk 5 above)
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/UserUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/at-hieucap/java/vn/asiantech/internship/models/Login.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    public void setPassword(String password) {
+        this.password = password;
+    }
+}
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/models/Login.java]: break 1 line

File: app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    private EditText mEdtUserName;
+    private EditText mEdtPassword;
+    private Button mBtnLogin;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_unit_test);
+        initViews();
+        setButtonLogin();
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java]: `setListener();`

File: app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        mBtnLogin.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                validate();
+            }
+        });
+    }
+
+    private void validate() {
+        if (validateUserName(mEdtUserName.getText().toString()) &&
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java]: `mEdtUserName.getText().toString().trim();` is better

File: app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+
+    private void initViews() {
+        mEdtUserName = findViewById(R.id.edtUserName);
+        mEdtPassword = findViewById(R.id.edtPassword);
+        mBtnLogin = findViewById(R.id.btnLogin);
+    }
+
+    private boolean validateUserName(String userName) {
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java]: name method should be `isValidateUserName`

File: app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        boolean validationUserName = false;
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java]: boolean variable start by `is`

File: app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        } else if (!UserNameValidation.checkHaveSpace(userName)) {
+            showToast(getString(R.string.test_user_name_have_space));
+        } else if (!UserNameValidation.checkCharacters(userName)) {
+            showToast(getString(R.string.test_user_name_special_character));
+        } else if (!UserNameValidation.checkDigitNumber(userName)) {
+            showToast(getString(R.string.test_user_name_digit_number));
+        } else if (!UserNameValidation.checkUpperCaseOrLowerCase(userName)) {
+            showToast(getString(R.string.test_user_name_upper_case_or_lower_case));
+        } else {
+            validationUserName = true;
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java]: `return true;`

File: app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        return validationUserName;
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java]: `return false;` you no need define `validationUserName`

File: app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    private boolean validatePassWord(String password) {
+        boolean validationPassword = false;
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java]: no need this variable

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
+package vn.asiantech.internship.unit_test;
+
+import java.util.regex.Pattern;
+
+/**
+ * Create check password
+ * Created by tiboo on 03/01/2018.
+ */
+public final class PasswordValidation {
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java]: add `private constructer`

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    public static boolean checkDifferentUserName(String userName, String password) {
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java]: `isDifferentUserName` is better

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        Boolean isDigitNumber = (Pattern.compile("[0-9 ]")).matcher(password).find();
+        return isCharacterSpecial || isDigitNumber;
+    }
+
+    public static boolean checkLength(String password) {
+        return password.length() > 7;
+    }
+
+    public static boolean checkSameCharacter(String password) {
+        boolean isSame = false;
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java]: no need define `variable`

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+ * Create check password
+ * Created by tiboo on 03/01/2018.
+ */
+public final class PasswordValidation {
+    public static boolean checkDifferentUserName(String userName, String password) {
+        return !password.equals(userName);
+    }
+
+    public static boolean checkCharacterSpecialAndDigitNumber(String password) {
+        Boolean isCharacterSpecial = password.matches("[a-zA-Z.? ]*");
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java]: no need define

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+    public static boolean checkSameCharacter(String password) {
+        boolean isSame = false;
+        int count = 0;
+        String[] text = password.split("");
+        for (int i = 0; i < text.length; i++) {
+            for (int j = i + 1; j < text.length; j++) {
+                if (text[i].equals(text[j])) {
+                    count++;
+                    if (count <= 2) {
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java]: move to out site `for`

Comment (at-trungnguyen) [reply app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java]: wrong condition

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        }
+        return isSame;
+    }
+
+    public static boolean checkHaveSpace(String password) {
+        return !password.contains(" ");
+    }
+
+    public static boolean checkCapitalLetter(String password) {
+        char character;
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java]: no need variables `character`, `capitalFlag`

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+                    if (count <= 2) {
+                        isSame = true;
+                    }
+                }
+            }
+        }
+        return isSame;
+    }
+
+    public static boolean checkHaveSpace(String password) {
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidation.java]: method return `boolean` should start by `is`

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UserNameValidation.java [hunk 21]
```diff
@@ -0,0 +1,51 @@
+package vn.asiantech.internship.unit_test;
+
+import java.util.Locale;
+
+/**
+ * Create check user name
+ * Created by tiboo on 03/01/2018.
+ */
+public final class UserNameValidation {
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UserNameValidation.java]: fix class same above

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UserNameValidation.java [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    public static boolean checkLength(String userName) {
+        return userName.length() > 5 && userName.length() < 24;
+    }
+
+    public static boolean checkCapitalLetter(String userName) {
+        char ch;
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UserNameValidation.java]: no need define

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UserNameValidation.java (hunk 21 above)
Comment (at-trungnguyen) [reply app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UserNameValidation.java]: name method is not good

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UserNameValidation.java [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    public static boolean checkHaveSpace(String userName) {
+        return !userName.contains(" ");
+    }
+
+    public static boolean checkCharacters(String userName) {
+        return userName.matches("[A-Za-z0-9]+");
+    }
+
+    public static boolean checkDigitNumber(String userName) {
+        char ch;
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UserNameValidation.java]: reduce variable is better

File: app/src/at-hieucap/res/layout/activity_unit_test.xml [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+    <EditText
+        android:id="@+id/edtPassword"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_below="@+id/edtUserName"
+        android:layout_marginTop="50dp"
+        android:gravity="center_vertical"
+        android:hint="@string/password"
+        android:textSize="25dp"
+        tools:ignore="TextFields" />
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/res/layout/activity_unit_test.xml]: please remove it.

File: app/src/testAt-hieucapDebug/java/UserUnitTest.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+    @Test
+    public void testHaveSpace() {
+        when(mLogIn.getUser()).thenReturn(TEXT_TEST);
+        Assert.assertTrue(UserNameValidation.checkHaveSpace(mLogIn.getUser()));
+    }
+
+    @Test
+    public void testDigitNumber() {
+        when(mLogIn.getUser()).thenReturn(TEXT_TEST);
+
```
Comment (at-trungnguyen) [inline app/src/testAt-hieucapDebug/java/UserUnitTest.java]: remove this line

File: app/src/testAt-hieucapDebug/java/PasswordUnitTest.java [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+import vn.asiantech.internship.unit_test.PasswordValidationTest;
+
+import static org.mockito.Mockito.when;
+
+/**
+ * Test validation password
+ * Created by tiboo on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PasswordUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/PasswordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-hieucapDebug/java/UsernameUnitTest.java [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+import vn.asiantech.internship.unit_test.UsernameValidationTest;
+
+import static org.mockito.Mockito.when;
+
+/**
+ * Test user name
+ * Created by tiboo on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class UsernameUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/UsernameUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

**Comment by at-hieucap** [comment]
@at-trungnguyen I fished exercise , please review for me,
Thank you so much !

File: app/src/testAt-hieucapDebug/java/PasswordUnitTest.java (hunk 26 above)
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/PasswordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-hieucapDebug/java/UsernameUnitTest.java (hunk 27 above)
Comment (at-dev-support-01) [inline app/src/testAt-hieucapDebug/java/UsernameUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+        mBtnLogin.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                validate();
+            }
+        });
+    }
+
+    private void validate() {
+        if (isValidateUserName(mEdtUserName.getText().toString()) &&
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/ui/mockito/LoginActivity.java]: `mEdtUserName.getText().toString().trim()` 🗡

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidationTest.java [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.regex.Pattern;
+
+/**
+ * Create check password
+ * Created by tiboo on 03/01/2018.
+ */
+public final class PasswordValidationTest {
+    private PasswordValidationTest() {
+        // No-oop
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/PasswordValidationTest.java]: `// No-oop` or `// No-op` 🗡

File: app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UsernameValidationTest.java [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.util.Locale;
+
+/**
+ * Create check user name
+ * Created by tiboo on 03/01/2018.
+ */
+public final class UsernameValidationTest {
+    private UsernameValidationTest() {
+        // No-oop
```
Comment (at-trungnguyen) [inline app/src/at-hieucap/java/vn/asiantech/internship/unit_test/UsernameValidationTest.java]: I think it's `// No-op`

**Comment by at-hieucap** [comment]
@at-trungnguyen I fished exercise , please review for me ,
Thank you so much !

**Comment by at-trungnguyen** [comment]
@at-hieucap Thanks, LGTM!

**Pull request merged**